File: src/at/bitfire/davdroid/resource/DavResourceFinder.java
package at.bitfire.davdroid.resource;

import java.io.Closeable;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.LinkedList;
import java.util.List;

import org.xbill.DNS.Lookup;
import org.xbill.DNS.Record;
import org.xbill.DNS.SRVRecord;
import org.xbill.DNS.TXTRecord;
import org.xbill.DNS.TextParseException;
import org.xbill.DNS.Type;

import android.content.Context;
import android.util.Log;
import at.bitfire.davdroid.R;
import at.bitfire.davdroid.webdav.DavException;
import at.bitfire.davdroid.webdav.DavHttpClient;
import at.bitfire.davdroid.webdav.DavIncapableException;
import at.bitfire.davdroid.webdav.HttpPropfind.Mode;
import at.bitfire.davdroid.webdav.NotAuthorizedException;
import at.bitfire.davdroid.webdav.WebDavResource;
import ch.boye.httpclientandroidlib.HttpException;
import ch.boye.httpclientandroidlib.impl.client.CloseableHttpClient;
import ezvcard.VCardVersion;

public class DavResourceFinder implements Closeable {
	private final static String TAG = "davdroid.DavResourceFinder";
	
	protected Context context;
	protected CloseableHttpClient httpClient;
	
	
	public DavResourceFinder(Context context) {
		this.context = context;
		
		// disable compression and enable network logging for debugging purposes 
		httpClient = DavHttpClient.create(true, true);
	}

	@Override
	public void close() throws IOException {
		httpClient.close();
	}
	
	
	public void findResources(ServerInfo serverInfo) throws URISyntaxException, DavException, HttpException, IOException {
		// CardDAV
		WebDavResource principal = getCurrentUserPrincipal(serverInfo, "carddav");
		if (principal != null) {
			serverInfo.setCardDAV(true);
		
			principal.propfind(Mode.HOME_SETS);
			String pathAddressBooks = principal.getAddressbookHomeSet();
			if (pathAddressBooks != null) {
				Log.i(TAG, "Found address book home set: " + pathAddressBooks);
			
				WebDavResource homeSetAddressBooks = new WebDavResource(principal, pathAddressBooks);
				if (checkHomesetCapabilities(homeSetAddressBooks, "addressbook")) {
					homeSetAddressBooks.propfind(Mode.CARDDAV_COLLECTIONS);
					
					List<ServerInfo.ResourceInfo> addressBooks = new LinkedList<ServerInfo.ResourceInfo>();
					if (homeSetAddressBooks.getMembers() != null)
						for (WebDavResource resource : homeSetAddressBooks.getMembers())
							if (resource.isAddressBook()) {
								Log.i(TAG, "Found address book: " + resource.getLocation().getPath());
								ServerInfo.ResourceInfo info = new ServerInfo.ResourceInfo(
									ServerInfo.ResourceInfo.Type.ADDRESS_BOOK,
									resource.isReadOnly(),
									resource.getLocation().toString(),
									resource.getDisplayName(),
									resource.getDescription(), resource.getColor()
								);
								
								VCardVersion version = resource.getVCardVersion();
								if (version == null)
									version = VCardVersion.V3_0;	// VCard 3.0 MUST be supported
								info.setVCardVersion(version);
								
								addressBooks.add(info);
							}
					serverInfo.setAddressBooks(addressBooks);
				} else
					Log.w(TAG, "Found address-book home set, but it doesn't advertise CardDAV support");
			}
		}
		
		// CalDAV
		principal = getCurrentUserPrincipal(serverInfo, "caldav");
		if (principal != null) {
			serverInfo.setCalDAV(true);

			principal.propfind(Mode.HOME_SETS);
			String pathCalendars = principal.getCalendarHomeSet();
			if (pathCalendars != null) {
				Log.i(TAG, "Found calendar home set: " + pathCalendars);
			
				WebDavResource homeSetCalendars = new WebDavResource(principal, pathCalendars);
				if (checkHomesetCapabilities(homeSetCalendars, "calendar-access")) {
					homeSetCalendars.propfind(Mode.CALDAV_COLLECTIONS);
					
					List<ServerInfo.ResourceInfo> calendars = new LinkedList<ServerInfo.ResourceInfo>();
					if (homeSetCalendars.getMembers() != null)
						for (WebDavResource resource : homeSetCalendars.getMembers())
							if (resource.isCalendar()) {
								Log.i(TAG, "Found calendar: " + resource.getLocation().getPath());
								if (resource.getSupportedComponents() != null) {
									// CALDAV:supported-calendar-component-set available
									boolean supportsEvents = false;
									for (String supportedComponent : resource.getSupportedComponents())
										if (supportedComponent.equalsIgnoreCase("VEVENT"))
											supportsEvents = true;
									if (!supportsEvents) {	// ignore collections without VEVENT support
										Log.i(TAG, "Ignoring this calendar because of missing VEVENT support");
										continue;
									}
								}
								ServerInfo.ResourceInfo info = new ServerInfo.ResourceInfo(
									ServerInfo.ResourceInfo.Type.CALENDAR,
									resource.isReadOnly(),
									resource.getLocation().toString(),
									resource.getDisplayName(),
									resource.getDescription(), resource.getColor()
								);
								info.setTimezone(resource.getTimezone());
								calendars.add(info);
							}
					serverInfo.setCalendars(calendars);
				} else
					Log.w(TAG, "Found calendar home set, but it doesn't advertise CalDAV support");
			}
		}
						
		if (!serverInfo.isCalDAV() && !serverInfo.isCardDAV())
			throw new DavIncapableException(context.getString(R.string.neither_caldav_nor_carddav));

	}
	
	
	/**
	 * Finds the initial service URL from a given base URI (HTTP[S] or mailto URI, user name, password)
	 * @param serverInfo	User-given service information (including base URI, i.e. HTTP[S] URL+user name+password or mailto URI and password)
	 * @param serviceName	Service name ("carddav" or "caldav")
	 * @return				Initial service URL (HTTP/HTTPS), without user credentials
	 * @throws URISyntaxException when the user-given URI is invalid
	 * @throws MalformedURLException when the user-given URI is invalid
	 * @throws UnknownServiceURLException when no intial service URL could be determined
	 */
	URL getInitialURL(ServerInfo serverInfo, String serviceName) throws URISyntaxException, MalformedURLException {
		String	scheme = null,
				domain = null;
		int		port = -1;
		String	path = "/";
		
		URI baseURI = serverInfo.getBaseURI();
		if ("mailto".equalsIgnoreCase(baseURI.getScheme())) {
			// mailto URIs
			String mailbox = serverInfo.getBaseURI().getSchemeSpecificPart();

			// determine service FQDN
			int pos = mailbox.lastIndexOf("@");
			if (pos == -1)
				throw new URISyntaxException(mailbox, "Email address doesn't contain @");
			domain = mailbox.substring(pos + 1);
		} else {
			// HTTP(S) URLs
			scheme = baseURI.getScheme();
			domain = baseURI.getHost();
			port = baseURI.getPort();
			path = baseURI.getPath();
		}

		// try to determine FQDN and port number using SRV records
		try {
			String name = "_" + serviceName + "s._tcp." + domain;
			Log.d(TAG, "Looking up SRV records for " + name);
			Record[] records = new Lookup(name, Type.SRV).run();
			if (records != null && records.length >= 1) {
				SRVRecord srv = selectSRVRecord(records);
				
				scheme = "https";
				domain = srv.getTarget().toString(true);
				port = srv.getPort();
				Log.d(TAG, "Found " + serviceName + "s service for " + domain + " -> " + domain + ":" + port);

				// SRV record found, look for TXT record too (for initial context path)
				records = new Lookup(name, Type.TXT).run();
				if (records != null && records.length >= 1) {
					TXTRecord txt = (TXTRecord)records[0];
					for (String segment : (String[])txt.getStrings().toArray())
						if (segment.startsWith("path=")) {
							path = segment.substring(5);
							Log.d(TAG, "Found initial context path for " + serviceName + " at " + domain + " -> " + path);
							break;
						}
				}
			}
		} catch (TextParseException e) {
			throw new URISyntaxException(domain, "Invalid domain name");
		}
		
		if (port != -1)
			return new URL(scheme, domain, port, path);
		else
			return new URL(scheme, domain, path);
	}
	
	
	/**
	 * Detects the current-user-principal for a given WebDavResource. At first, /.well-known/ is tried. Only
	 * if no current-user-principal can be detected for the .well-known location, the given location of the resource
	 * is tried.
	 * @param resource 		Location that will be queried
	 * @param serviceName	Well-known service name ("carddav", "caldav")
	 * @return	WebDavResource of current-user-principal for the given service, or null if it can't be found
	 */
	WebDavResource getCurrentUserPrincipal(ServerInfo serverInfo, String serviceName) throws URISyntaxException, IOException, NotAuthorizedException {
		URL initialURL = getInitialURL(serverInfo, serviceName);
		
		// determine base URL (host name and initial context path)
		WebDavResource base = new WebDavResource(httpClient,
				//new URI(URIUtils.ensureTrailingSlash(serverInfo.getBaseURI())),
				initialURL,
				serverInfo.getUserName(), serverInfo.getPassword(), serverInfo.isAuthPreemptive());
		
		// look for well-known service (RFC 5785)
		try {
			WebDavResource wellKnown = new WebDavResource(base, "/.well-known/" + serviceName);
			wellKnown.propfind(Mode.CURRENT_USER_PRINCIPAL);
			if (wellKnown.getCurrentUserPrincipal() != null)
				return new WebDavResource(wellKnown, wellKnown.getCurrentUserPrincipal());
		} catch (NotAuthorizedException e) {
			Log.d(TAG, "Well-known " + serviceName + " service detection not authorized", e);
			throw e;
		} catch (HttpException e) {
			Log.d(TAG, "Well-known " + serviceName + " service detection failed with HTTP error", e);
		} catch (DavException e) {
			Log.d(TAG, "Well-known " + serviceName + " service detection failed at DAV level", e);
		}

		// fall back to user-given initial context path
		try {
			base.propfind(Mode.CURRENT_USER_PRINCIPAL);
			if (base.getCurrentUserPrincipal() != null)
				return new WebDavResource(base, base.getCurrentUserPrincipal());
		} catch (NotAuthorizedException e) {
			Log.d(TAG, "Not authorized for querying principal for " + serviceName + " service", e);
			throw e;
		} catch (HttpException e) {
			Log.d(TAG, "HTTP error when querying principal for " + serviceName + " service", e);
		} catch (DavException e) {
			Log.d(TAG, "DAV error when querying principal for " + serviceName + " service", e);
		}
		Log.i(TAG, "Couldn't find current-user-principal for service " + serviceName);
		return null;
	}
	
	private static boolean checkHomesetCapabilities(WebDavResource resource, String davCapability) throws URISyntaxException, IOException {
		// check for necessary capabilities
		try {
			resource.options();
			if (resource.supportsDAV(davCapability) &&
				resource.supportsMethod("PROPFIND"))		// check only for methods that MUST be available for home sets
				return true;
		} catch(HttpException e) {
			// for instance, 405 Method not allowed
		}
		return false;
	}
	
	
	SRVRecord selectSRVRecord(Record[] records) {
		if (records.length > 1)
			Log.w(TAG, "Multiple SRV records not supported yet; using first one");
		return (SRVRecord)records[0];
	}

}

File: app/src/main/kotlin/at/bitfire/davdroid/servicedetection/DavResourceFinder.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/
package at.bitfire.davdroid.servicedetection

import android.content.Context
import at.bitfire.dav4jvm.DavResource
import at.bitfire.dav4jvm.Property
import at.bitfire.dav4jvm.Response
import at.bitfire.dav4jvm.UrlUtils
import at.bitfire.dav4jvm.exception.DavException
import at.bitfire.dav4jvm.exception.HttpException
import at.bitfire.dav4jvm.exception.UnauthorizedException
import at.bitfire.dav4jvm.property.*
import at.bitfire.dav4jvm.property.caldav.CalendarColor
import at.bitfire.dav4jvm.property.caldav.CalendarDescription
import at.bitfire.dav4jvm.property.caldav.CalendarHomeSet
import at.bitfire.dav4jvm.property.caldav.CalendarTimezone
import at.bitfire.dav4jvm.property.caldav.CalendarUserAddressSet
import at.bitfire.dav4jvm.property.caldav.SupportedCalendarComponentSet
import at.bitfire.dav4jvm.property.carddav.AddressbookDescription
import at.bitfire.dav4jvm.property.carddav.AddressbookHomeSet
import at.bitfire.dav4jvm.property.webdav.CurrentUserPrincipal
import at.bitfire.dav4jvm.property.webdav.CurrentUserPrivilegeSet
import at.bitfire.dav4jvm.property.webdav.DisplayName
import at.bitfire.dav4jvm.property.webdav.HrefListProperty
import at.bitfire.dav4jvm.property.webdav.ResourceType
import at.bitfire.davdroid.db.Collection
import at.bitfire.davdroid.db.Credentials
import at.bitfire.davdroid.log.StringHandler
import at.bitfire.davdroid.network.HttpClient
import at.bitfire.davdroid.util.DavUtils
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.apache.commons.lang3.builder.ReflectionToStringBuilder
import org.xbill.DNS.Lookup
import org.xbill.DNS.Type
import java.io.InterruptedIOException
import java.net.SocketTimeoutException
import java.net.URI
import java.net.URISyntaxException
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Does initial resource detection (straight after app install). Called after user has supplied url in
 * app setup process [at.bitfire.davdroid.ui.setup.DetectConfigurationFragment].
 * It uses the (user given) base URL to find
 * - services (CalDAV and/or CardDAV),
 * - principal,
 * - homeset/collections (multistatus responses are handled through dav4jvm).
 *
 * @param context        to build the HTTP client
 * @param baseURI        user-given base URI (either mailto: URI or http(s):// URL)
 * @param credentials    optional login credentials (username/password, client certificate, OAuth state)
 */
class DavResourceFinder(
    val context: Context,
    private val baseURI: URI,
    private val credentials: Credentials? = null
): AutoCloseable {

    enum class Service(val wellKnownName: String) {
        CALDAV("caldav"),
        CARDDAV("carddav");

        override fun toString() = wellKnownName
    }

    val log: Logger = Logger.getLogger("davx5.DavResourceFinder")
    private val logBuffer = StringHandler()
    init {
        log.level = Level.FINEST
        log.addHandler(logBuffer)
    }

    var encountered401 = false

    private val httpClient: HttpClient = HttpClient.Builder(context, logger = log).let {
        credentials?.let { credentials ->
            it.addAuthentication(null, credentials)
        }
        it.setForeground(true)
        it.build()
    }

    override fun close() {
        httpClient.close()
    }


    /**
     * Finds the initial configuration (= runs the service detection process).
     *
     * In case of an error, it returns an empty [Configuration] with error logs
     * instead of throwing an [Exception].
     *
     * @return service information – if there's neither a CalDAV service nor a CardDAV service,
     * service detection was not successful
     */
    fun findInitialConfiguration(): Configuration {
        var cardDavConfig: Configuration.ServiceInfo? = null
        var calDavConfig: Configuration.ServiceInfo? = null

        try {
            try {
                cardDavConfig = findInitialConfiguration(Service.CARDDAV)
            } catch (e: Exception) {
                log.log(Level.INFO, "CardDAV service detection failed", e)
                processException(e)
            }

            try {
                calDavConfig = findInitialConfiguration(Service.CALDAV)
            } catch (e: Exception) {
                log.log(Level.INFO, "CalDAV service detection failed", e)
                processException(e)
            }
        } catch(e: Exception) {
            // we have been interrupted; reset results so that an error message will be shown
            cardDavConfig = null
            calDavConfig = null
        }

        return Configuration(
            cardDAV = cardDavConfig,
            calDAV = calDavConfig,
            encountered401 = encountered401,
            logs = logBuffer.toString()
        )
    }

    private fun findInitialConfiguration(service: Service): Configuration.ServiceInfo? {
        // domain for service discovery
        var discoveryFQDN: String? = null

        // discovered information goes into this config
        val config = Configuration.ServiceInfo()

        // Start discovering
        log.info("Finding initial ${service.wellKnownName} service configuration")
        when (baseURI.scheme.lowercase()) {
            "http", "https" ->
                baseURI.toHttpUrlOrNull()?.let { baseURL ->
                    // remember domain for service discovery
                    if (baseURL.scheme.equals("https", true))
                        // service discovery will only be tried for https URLs, because only secure service discovery is implemented
                        discoveryFQDN = baseURL.host

                    // Actual discovery process
                    checkBaseURL(baseURL, service, config)

                    // If principal was not found already, try well known URI
                    if (config.principal == null)
                        try {
                            config.principal = getCurrentUserPrincipal(baseURL.resolve("/.well-known/" + service.wellKnownName)!!, service)
                        } catch(e: Exception) {
                            log.log(Level.FINE, "Well-known URL detection failed", e)
                            processException(e)
                        }
                }
            "mailto" -> {
                val mailbox = baseURI.schemeSpecificPart
                val posAt = mailbox.lastIndexOf("@")
                if (posAt != -1)
                    discoveryFQDN = mailbox.substring(posAt + 1)
            }
        }

        // Second try: If user-given URL didn't reveal a principal, search for it (SERVICE DISCOVERY)
        if (config.principal == null)
            discoveryFQDN?.let { fqdn ->
                log.info("No principal found at user-given URL, trying to discover for domain $fqdn")
                try {
                    config.principal = discoverPrincipalUrl(fqdn, service)
                } catch(e: Exception) {
                    log.log(Level.FINE, "$service service discovery failed", e)
                    processException(e)
                }
            }

        // detect email address
        if (service == Service.CALDAV)
            config.principal?.let { principal ->
                config.emails.addAll(queryEmailAddress(principal))
            }

        // return config or null if config doesn't contain useful information
        val serviceAvailable = config.principal != null || config.homeSets.isNotEmpty() || config.collections.isNotEmpty()
        return if (serviceAvailable)
            config
        else
            null
    }

    /**
     * Entry point of the actual discovery process.
     *
     * Queries the user-given URL (= base URL) to detect whether it contains a current-user-principal
     * or whether it is a homeset or collection.
     *
     * @param baseURL   base URL provided by the user
     * @param service   service to detect configuration for
     * @param config    found configuration will be written to this object
     */
    private fun checkBaseURL(baseURL: HttpUrl, service: Service, config: Configuration.ServiceInfo) {
        log.info("Checking user-given URL: $baseURL")

        val davBaseURL = DavResource(httpClient.okHttpClient, baseURL, log)
        try {
            when (service) {
                Service.CARDDAV -> {
                    davBaseURL.propfind(
                        0,
                        ResourceType.NAME, DisplayName.NAME, AddressbookDescription.NAME,
                        AddressbookHomeSet.NAME,
                        CurrentUserPrincipal.NAME
                    ) { response, _ ->
                        scanResponse(ResourceType.ADDRESSBOOK, response, config)
                    }
                }
                Service.CALDAV -> {
                    davBaseURL.propfind(
                        0,
                        ResourceType.NAME, DisplayName.NAME, CalendarColor.NAME, CalendarDescription.NAME, CalendarTimezone.NAME, CurrentUserPrivilegeSet.NAME, SupportedCalendarComponentSet.NAME,
                        CalendarHomeSet.NAME,
                        CurrentUserPrincipal.NAME
                    ) { response, _ ->
                        scanResponse(ResourceType.CALENDAR, response, config)
                    }
                }
            }
        } catch(e: Exception) {
            log.log(Level.FINE, "PROPFIND/OPTIONS on user-given URL failed", e)
            processException(e)
        }
    }

    /**
     * Queries a user's email address using CalDAV scheduling: calendar-user-address-set.
     * @param principal principal URL of the user
     * @return list of found email addresses (empty if none)
     */
    fun queryEmailAddress(principal: HttpUrl): List<String> {
        val mailboxes = LinkedList<String>()
        try {
            DavResource(httpClient.okHttpClient, principal, log).propfind(0, CalendarUserAddressSet.NAME) { response, _ ->
                response[CalendarUserAddressSet::class.java]?.let { addressSet ->
                    for (href in addressSet.hrefs)
                        try {
                            val uri = URI(href)
                            if (uri.scheme.equals("mailto", true))
                                mailboxes.add(uri.schemeSpecificPart)
                        } catch(e: URISyntaxException) {
                            log.log(Level.WARNING, "Couldn't parse user address", e)
                        }
                }
            }
        } catch(e: Exception) {
            log.log(Level.WARNING, "Couldn't query user email address", e)
            processException(e)
        }
        return mailboxes
    }

    /**
     * Depending on [resourceType] (CalDAV or CardDAV), this method checks whether [davResponse] references
     * - an address book or calendar (actual resource), and/or
     * - an "address book home set" or a "calendar home set", and/or
     * - whether it's a principal.
     *
     * Respectively, this method will add the response to [config.collections], [config.homesets] and/or [config.principal].
     * Collection URLs will be stored with trailing "/".
     *
     * @param resourceType  type of service to search for in the response
     * @param davResponse   response whose properties are evaluated
     * @param config        structure storing the references
     */
    fun scanResponse(resourceType: Property.Name, davResponse: Response, config: Configuration.ServiceInfo) {
        var principal: HttpUrl? = null

        // Type mapping
        val homeSetClass: Class<out HrefListProperty>
        val serviceType: Service
        when (resourceType) {
            ResourceType.ADDRESSBOOK -> {
                homeSetClass = AddressbookHomeSet::class.java
                serviceType = Service.CARDDAV
            }
            ResourceType.CALENDAR -> {
                homeSetClass = CalendarHomeSet::class.java
                serviceType = Service.CALDAV
            }
            else -> throw IllegalArgumentException()
        }

        // check for current-user-principal
        davResponse[CurrentUserPrincipal::class.java]?.href?.let { currentUserPrincipal ->
            principal = davResponse.requestedUrl.resolve(currentUserPrincipal)
        }

        davResponse[ResourceType::class.java]?.let {
            // Is it a calendar or an address book, ...
            if (it.types.contains(resourceType))
                Collection.fromDavResponse(davResponse)?.let { info ->
                    log.info("Found resource of type $resourceType at ${info.url}")
                    config.collections[info.url] = info
                }

            // ... and/or a principal?
            if (it.types.contains(ResourceType.PRINCIPAL))
                principal = davResponse.href
        }

        // Is it an addressbook-home-set or calendar-home-set?
        davResponse[homeSetClass]?.let { homeSet ->
            for (href in homeSet.hrefs) {
                davResponse.requestedUrl.resolve(href)?.let {
                    val location = UrlUtils.withTrailingSlash(it)
                    log.info("Found home-set of type $resourceType at $location")
                    config.homeSets += location
                }
            }
        }

        // Is there a principal too?
        principal?.let {
            if (providesService(it, serviceType))
                config.principal = principal
            else
                log.warning("Principal $principal doesn't provide $serviceType service")
        }
    }

    /**
     * Sends an OPTIONS request to determine whether a URL provides a given service.
     *
     * @param url      URL to check; often a principal URL
     * @param service  service to check for
     *
     * @return whether the URL provides the given service
     */
    fun providesService(url: HttpUrl, service: Service): Boolean {
        var provided = false
        try {
            DavResource(httpClient.okHttpClient, url, log).options { capabilities, _ ->
                if ((service == Service.CARDDAV && capabilities.contains("addressbook")) ||
                    (service == Service.CALDAV && capabilities.contains("calendar-access")))
                    provided = true
            }
        } catch(e: Exception) {
            log.log(Level.SEVERE, "Couldn't detect services on $url", e)
            if (e !is HttpException && e !is DavException)
                throw e
        }
        return provided
    }


    /**
     * Try to find the principal URL by performing service discovery on a given domain name.
     * Only secure services (caldavs, carddavs) will be discovered!
     *
     * @param domain         domain name, e.g. "icloud.com"
     * @param service        service to discover (CALDAV or CARDDAV)
     * @return principal URL, or null if none found
     */
    fun discoverPrincipalUrl(domain: String, service: Service): HttpUrl? {
        val scheme: String
        val fqdn: String
        var port = 443
        val paths = LinkedList<String>()     // there may be multiple paths to try

        val query = "_${service.wellKnownName}s._tcp.$domain"
        log.fine("Looking up SRV records for $query")

        val srvLookup = Lookup(query, Type.SRV)
        DavUtils.prepareLookup(context, srvLookup)
        val srv = DavUtils.selectSRVRecord(srvLookup.run().orEmpty())

        if (srv != null) {
            // choose SRV record to use (query may return multiple SRV records)
            scheme = "https"
            fqdn = srv.target.toString(true)
            port = srv.port
            log.info("Found $service service at https://$fqdn:$port")
        } else {
            // no SRV records, try domain name as FQDN
            log.info("Didn't find $service service, trying at https://$domain:$port")

            scheme = "https"
            fqdn = domain
        }

        // look for TXT record too (for initial context path)
        val txtLookup = Lookup(query, Type.TXT)
        DavUtils.prepareLookup(context, txtLookup)
        paths.addAll(DavUtils.pathsFromTXTRecords(txtLookup.run()))

        // in case there's a TXT record, but it's wrong, try well-known
        paths.add("/.well-known/" + service.wellKnownName)
        // if this fails too, try "/"
        paths.add("/")

        for (path in paths)
            try {
                val initialContextPath = HttpUrl.Builder()
                        .scheme(scheme)
                        .host(fqdn).port(port)
                        .encodedPath(path)
                        .build()

                log.info("Trying to determine principal from initial context path=$initialContextPath")
                val principal = getCurrentUserPrincipal(initialContextPath, service)

                principal?.let { return it }
            } catch(e: Exception) {
                log.log(Level.WARNING, "No resource found", e)
                processException(e)
            }
        return null
    }

    /**
     * Queries a given URL for current-user-principal
     *
     * @param url       URL to query with PROPFIND (Depth: 0)
     * @param service   required service (may be null, in which case no service check is done)
     * @return          current-user-principal URL that provides required service, or null if none
     */
    fun getCurrentUserPrincipal(url: HttpUrl, service: Service?): HttpUrl? {
        var principal: HttpUrl? = null
        DavResource(httpClient.okHttpClient, url, log).propfind(0, CurrentUserPrincipal.NAME) { response, _ ->
            response[CurrentUserPrincipal::class.java]?.href?.let { href ->
                response.requestedUrl.resolve(href)?.let {
                    log.info("Found current-user-principal: $it")

                    // service check
                    if (service != null && !providesService(it, service))
                        log.warning("Principal $it doesn't provide $service service")
                    else
                        principal = it
                }
            }
        }
        return principal
    }

    /**
     * Processes a thrown exception like this:
     *
     *   - If the Exception is an [UnauthorizedException] (HTTP 401), [encountered401] is set to *true*.
     *   - Re-throws the exception if it signals that the current thread was interrupted to stop the current operation.
     */
    private fun processException(e: Exception) {
        if (e is UnauthorizedException)
            encountered401 = true
        else if ((e is InterruptedIOException && e !is SocketTimeoutException) || e is InterruptedException)
            throw e
    }


    // data classes

    class Configuration(
        val cardDAV: ServiceInfo?,
        val calDAV: ServiceInfo?,

        val encountered401: Boolean,
        val logs: String
    ) {

        data class ServiceInfo(
            var principal: HttpUrl? = null,
            val homeSets: MutableSet<HttpUrl> = HashSet(),
            val collections: MutableMap<HttpUrl, Collection> = HashMap(),

            val emails: MutableList<String> = LinkedList()
        )

        override fun toString(): String {
            val builder = ReflectionToStringBuilder(this)
            builder.setExcludeFieldNames("logs")
            return builder.toString()
        }

    }

}
